### ISSUE TITLE
Moving to Typescript

### ISSUE THREAD

**Issue by wojpawlik** [telegraf/telegraf#1074]
### Non-breaking

- [x] Merge/close all pull requests.
- [x] Move `.js` files to `src/`, `tsc`, add `{ "exports": { "./": "./lib/" } }` to `package.json`.
- [x] ~~Add stub files to avoid breaking for Node <12~~ Drop Node <12 compat
- [x] Change extension to `.ts` and add type annotations from `.d.ts` files.
  + [x] `router.ts`
  + [x] `telegram.ts`
  + [x] `context.ts`
  + [x] `composer.ts`
  + [x] `core/network/client.ts`
  + [x] `telegraf.ts`
- [x] Configure formatting/linting of `.ts` files.
- [ ] Move the main documentation to `typedoc` (#1205)
- [ ] Update examples (#1177, #1194)

### Help wanted

- [x] `scenes/`
- [x] `stage.ts`
- [x] `extra.ts` (#1076)
- [x] `session.ts` (#914)
- [x] Complete type annotations, `noImplicitAny`.
- [ ] Fix eslint warnings and https://github.com/telegraf/telegraf/search?q=ts-expect-error.

### Breaking

- [x] `index.ts` ~~(awaiting https://github.com/microsoft/TypeScript/issues/39464)~~
- [x] `mjs-entry`
- [x] `"exports": { "import": "./lib/index.mjs", "require": "./lib/index.js" }`
- [x] `markup.ts` (rewrite)
- [x] Start building `.d.ts` files from source.
- [ ] Resolve [related issues](https://github.com/telegraf/telegraf/labels/typings).
- [ ] (optional) [denoify](https://github.com/garronej/denoify/issues/13#issuecomment-689076929).

**Comment by wojpawlik** [comment]
`export=` makes it difficult if not impossible to export related interfaces, so it has to go (breaking).

**Comment by wojpawlik** [comment]
I'm done for now. Would appreciate help with the rest of the files, or green light for breaking changes.

**Comment by KnorpelSenf** [comment]
Are you planning to do a full migration to TS, that is, eventually dropping `allowJs`? If so, I have a few questions regarding this issue, sorted by importance, highest first.

1) There is a number of types in the `typings` directory and also you have annotated the library directly in some places. These definitions overlap sometimes and even are contradictory, e.g. for `composer.ts` the class members differ. You're not emitting any declaration files yet as by `tsconfig.json`. Where do the types in the `typings` directory come from and what is your general plan about the contents in this directory?
1) I can see that the types of the Telegram API stem from the `telegram-typings` package. I can also see that you extend these types by unions of string literal types, such as `UpdateType`. That makes a lot of sense to me. But why are you redeclaring all those interfaces like `InputMediaPhoto` exactly the same way as in the package? TypeScript merges the definitions, so it does not have any effect. Am I missing something?
1) A migration to TS allows for new patterns such as async/await. You're currently not making use of this (except for one place). Should this be changed along with the migration?
1) Any reason not to use `strict` mode? It already works.
1) You're specifying `typescript@beta` in your `package.json`. The ecosystem is not yet at a point that every tool supports this version, such as your ESLint plugin `@typescript-eslint/typescript-estree`. Are you planning on releasing it with a beta version of TS or is a full TS release still far away?
1) You're targeting `ES2019`, thus dropping support for Node 10.x. Is that intentional? I'm not sure how relevant/prevalent this version is.
1) You're explicitly avoiding to generate a `package-lock.json` file. The `typescript` package does not follow semver, so specifying semver ranges alone in the `package.json` file does not prevent some changing transitive dependency to break your build. I understand that there are a few disadvantages to committing this file, however, migrating to TS could bring in a new factor to the equation. Do you maybe want to reconsider this decision?

Sorry for having so many questions. I'd like to see what I can do to help you migrate to TS because I'm really lacking proper type annotations every now and then. Before working on a series of PRs, however, I wanted to make sure that you're not planning some major changes to the project structure and/or configuration.

**Comment by wojpawlik** [comment]
Hello, help always appreciated :) Please refrain from touching `src/telegraf.ts` until #1084.

> Are you planning [...] dropping `allowJs`?

Not by myself. I'll happily accept PRs with more types and disable `allowJs` after it's no longer necessary.

> 1\. Where do the types in the `typings` directory come from and what is your general plan about the contents in this directory?

They're written by hand. After https://github.com/microsoft/TypeScript/issues/39464, I'll create `src/index.ts`, move `typings/telegram-types.d.ts` to `src/`, remove the rest of [`typings/`](https://github.com/telegraf/telegraf/tree/cff8967f9ba083b267825c872282344c24e88d24/typings), `.gitignore`, and start generating them.

> 2\. why are you redeclaring all those interfaces like `InputMediaPhoto` exactly the same way as in the package?

I haven't noticed before. I agree that it's questionable.

> 3\. A migration to TS allows for new patterns such as async/await. [...] Should this be changed along with the migration?

I personally don't, unless I'm changing the adjacent lines anyway.

> 4\. Any reason not to use `strict` mode?

Some implicit `any`s I didn't feel like dealing with. I'll be happy to enable it once all non-`core` files are fully typed.

> 5\. Are you planning on releasing it with a beta version of TS [...]?

I'm not permitted to publish to npm. Releases are up to @dotcypress. I need the beta now just for:

https://github.com/telegraf/telegraf/blob/cff8967f9ba083b267825c872282344c24e88d24/src/types.ts#L29

> 6\. dropping support for Node 10.x. Is that intentional?

Yes, it'll be EOL in 3 months.

> 7\. You're explicitly avoiding to generate a `package-lock.json` file. The `typescript` package does not follow semver

Valid point; I'm just reluctant to go against @dotcypress, who made the choice years ago.

**Comment by KnorpelSenf** [comment]
@wojpawlik thank you for the thorough explanations!

> I'll create src/index.ts, move typings/telegram-types.d.ts to src/, remove the rest of typings/, .gitignore, and start generating them.

That sounds really good and it also means nothing needs to be done for point 2.

> Some implicit `any`s I didn't feel like dealing with.

I was able to compile the project just like that with `strict` mode enabled. This basically means we could technically enable it right away. I'll wait for #1084 and check that again, taking care of implicit `any`s where necessary.

> Yes, it'll be EOL in 3 months.

Good point, I was asking because you said

> (optional) Add stub files to avoid breaking for Node <12.

in the original description above, so I assumed it might be of importance. Node 11 could be affected, too, but even the Debian 9 repo ships Node 12, so I guess this affects no one really.

@dotcypress please have a look at the last point in my previous post. I'm happy with either decision but this should at least be reconsidered due to the changes. Thanks

**Comment by wojpawlik** [comment]
My mistake, Node 10 goes EOL at April, not October. I just needed some relatively modern target for `tsconfig.json`, and I happen to happily use ES2019 in some other Node 12 project. I'm open to changing it if Node 10 support is decided desirable, despite not supporting `package.json`'s `"exports"` property. What's the best TS target for Node 10?

**Comment by KnorpelSenf** [comment]
[node.green](https://node.green) lists ES2017. But I don't really know if it's even desirable to support that. Do people even use that version? Supporting old versions is more than TS transpiling, there'd also need to be polyfills for newly introduced functions and so on … maybe @dotcypress should decide this once again?

**Comment by KnorpelSenf** [comment]
`package.json` lists Node 8.

**Comment by wojpawlik** [comment]
Works in TS 3.9, TIL:
```ts
type Tail<T extends any[]> = ((...a: T) => any) extends (f: any, ...r: infer R) => any ? R : []
```

**Comment by MrMarble** [comment]
Currently (3.38.0) there's a slight problem when setting the state on a context (like the recommended way says). The TelegrafContext type does not contains a state property thus giving a typescript error

![imagen](https://user-images.githubusercontent.com/4268580/88476697-d8c40800-cf3a-11ea-8cc8-1f406c3735a3.png)

**Comment by ekifox** [comment]
```js
await ctx.replyWithPhoto(
            { source: 'map.jpg' },
            Extra.HTML().caption('some text here')
)
```

```(method) Extra.caption(caption: string): Extra
Argument of type 'Extra' is not assignable to parameter of type 'ExtraPhoto'.
  Types of property 'caption' are incompatible.
    Type '(caption: string) => Extra' is not assignable to type 'string'.ts(2345)
```

**Comment by wojpawlik** [comment]
@MrMarble already fixed on develop, will be published in 4.0.

@ekifox You need `Extra.HTML().caption('some text here').extra()`, see #1076.

<hr>

I have no more breaking changes to suggest, so I'm ready for _rough_ 4.0 anytime. Caveats:

1. `Extra` and `Markup` wouldn't be available, since we're still working on them (#1134),
2. `Update` and `Message` types require explicit `in` checks due to https://github.com/microsoft/TypeScript/issues/4742.

**Comment by KnorpelSenf** [comment]
@wojpawlik do you know what it takes until the 4.0 release can be considered, other than completing #1134?

**Comment by wojpawlik** [comment]
Only #1134.

**Comment by wojpawlik** [comment]
I'm ready for 4.0.0 release @dotcypress, but you might wanna wait for https://github.com/telegraf/telegraf/pull/1134#issuecomment-708249167.

**Comment by KnorpelSenf** [comment]
Please check out #1156

**Comment by KnorpelSenf** [comment]
The PRs #1158 #1159 #1160 are my last contributions for now, then I'll wait for the major release

**Comment by wojpawlik** [comment]
4.0.0 can be released anytime, with or without #1159 @dotcypress. Fix the linting issue @KnorpelSenf :)

**Comment by KnorpelSenf** [comment]
> Fix the linting issue @KnorpelSenf :)

done, #1159 passes linting again

**Comment by wojpawlik** [comment]
4.0 release is now blocked by #1173.

**Comment by dotcypress** [comment]
We need to update all examples first!

**Comment by KnorpelSenf** [comment]
@dotcypress Should we provide both TS and JS examples or should we rather migrate the existing ones to TS?

**Comment by wojpawlik** [comment]
I'll take care of examples. I'll leave them in JS where possible, and convert those that need type annotations to TS.

**Comment by wojpawlik** [comment]
https://github.com/microsoft/TypeScript/issues/30602#issuecomment-725011949 was a setback, I should commit or PR examples tomorrow.

@KnorpelSenf, feel free to try to make `Composer` take advantage of "unionization" of `typegram`, I deemed it impossible at the moment.

**Comment by KnorpelSenf** [comment]
> @KnorpelSenf, feel free to try to make `Composer` take advantage of "unionization" of `typegram`, I deemed it impossible at the moment.

@wojpawlik I agree, this is really difficult because of the whole `context.ts` type mapping magic and the contravariant middleware arguments. I'll think some more about it, maybe we can refactor `context.ts` such that this becomes simpler. We'll see

**Comment by wojpawlik** [comment]
https://github.com/telegraf/telegraf/blob/d2126819c47f8668fedb4f9e4fe17c69ac2ce723/docs/examples/scenes-bot.js#L1
https://github.com/telegraf/telegraf/blob/d2126819c47f8668fedb4f9e4fe17c69ac2ce723/docs/examples/wizard-bot.js#L1

**Comment by septs** [comment]
Next release date?

**Comment by wojpawlik** [comment]
#1194 and #1195 are currently blocking 4.0.0 release, they should take up to a few more days, you can help by reviewing.

**Comment by KnorpelSenf** [comment]
#1194 is complete unless changes are requested.

#1195 could be merged as-is (after review!), but supporting channel mode would also be nice. I'll look into that on the coming days.

**Comment by wojpawlik** [comment]
Migration complete enough, closing in favor of #1202 and #1205.
